### PR TITLE
feat: cli arg for file permission of socket

### LINKED ISSUE #1466: Add option to set unix socket permissions
Hello,

when using the --socket option, I can tell code-server which socket to use, but not the permissions. At the moment the default permissions are 0755, which means that only the user is able to write to the socket while it's world readable...

When running together with a web server, it'd be nice if it could be set to 0770 and giving the group name/id so that a common group between web server and code-server would be possible.

Something like:

--socket /var/run/code-server.sock,0770,user,group
--socket /var/run/code-server.sock,0770,,group

Also, the server doesn't clean up the socket when it goes down and on a restart it errors out with address already in use...

I'm using workarounds at the moment, but it would be better if code-server could take care of it on its own.

### PR DESCRIPTION
**Pull request by TechCiel**
Fixes #1466

Add a cli option `--socket-permission`, and set file mode accordingly when listening a socket.

### CODE REVIEW

**Review by jsjoeio — APPROVED** [review]
Very clean PR 👏 Thank you for adding tests as well!

I'm less familiar with socket-permissions so I'll defer review to @code-asher

File: src/node/app.ts [hunk 1]
```diff
@@ -31,6 +31,8 @@ const listen = (server: http.Server, { host, port, socket }: ListenOptions) => {
       server.off("error", reject)
       server.on("error", (err) => util.logError(logger, "http server error", err))
 
+      if (socket && permission) fs.chmod(socket, permission)
```
Comment (code-asher) [inline src/node/app.ts:34]: We might want to catch any errors and log them; Node can be a bit weird about uncaught promise rejections:

```suggestion
      if (socket && permission) {
        fs.chmod(socket, permission).catch((error) => {
          util.logError(logger, "socket chmod", error)
        })
      }
```

Although maybe should this be a hard exit.  If someone is setting permissions and those fail to set maybe we should be really noisy about it.  For example we could do:

```suggestion
      if (socket && permission) {
        fs.chmod(socket, permission).then(resolve).catch(reject)
      } else {
        resolve()
      }
```

File: src/node/app.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ import { disposer } from "./http"
 import { isNodeJSErrnoException } from "./util"
 import { handleUpgrade } from "./wsRouter"
 
-type ListenOptions = Pick<DefaultedArgs, "socket" | "port" | "host">
+type ListenOptions = Pick<DefaultedArgs, "socket-permission" | "socket" | "port" | "host">
```
Comment (code-asher) [inline src/node/app.ts:14]: I do not feel strongly either way but what do we think about `--socket-mode` to match things like systemd?

**Review by code-asher — CHANGES_REQUESTED** [review]
Thank you for the contribution!  This has been a long-standing one.

File: src/node/app.ts (hunk 1 above)
Comment (TechCiel) [reply src/node/app.ts]: have changed to log & reject

File: src/node/app.ts (hunk 2 above)
Comment (TechCiel) [reply src/node/app.ts]: have changed to --socket-mode

**Review by code-asher — APPROVED** [review]

**Pull request merged**